be/src/exprs/agg/approx_top_k.h

```diff
@@ -0,0 +1,466 @@
+            }
+            for (size_t i = start; i <= end; i++) {
+                table[counters[i].value] = &counters[i];
+            }
+        }
+        DCHECK(std::is_sorted(counters.begin(), counters.begin() + unused_idx, cmp));
+        [[maybe_unused]] auto check = [this]() -> bool {
+            // Check index
+            for (size_t i = 0; i < counter_num; i++) {
+                if (i != counters[i]._index) {
+                    return false;
+                }
+            }
+            // Check table
+            static EQ eq;
+            for (auto& [k, v] : table) {
+                if (!eq(k, v->value)) {
+                    return false;
+                }
+            }
+            return true;
+        };
+        DCHECK(check());
+    }
+
+    template <bool IsDeepCopy>
+    CppType _copy(MemPool* mem_pool, const CppType& value) {
+        if constexpr (IsDeepCopy && IsSlice<CppType>) {
+            uint8_t* pos = mem_pool->allocate(value.size);
+            std::memcpy(pos, value.data, value.size);
+            return Slice{pos, value.size};
+        } else {
+            return value;
+        }
+    }
+};
+
+template <LogicalType LT, typename T = RunTimeCppType<LT>>
+class ApproxTopKAggregateFunction final
+        : public AggregateFunctionBatchHelper<ApproxTopKState<LT>, ApproxTopKAggregateFunction<LT, T>> {
+public:
+    using CppType = RunTimeCppType<LT>;
+    using InputColumnType = RunTimeColumnType<LT>;
+
+    static constexpr int32_t DEFAULT_K = 5;
+    static constexpr int32_t MAX_COUNTER_NUM = 100000;
+
+    std::pair<int32_t, int32_t> get_k_and_counter_num(FunctionContext* ctx) const {
+        int32_t k = DEFAULT_K;
+        if (ctx->get_num_args() > 1) {
+            k = ColumnHelper::get_const_value<TYPE_INT>(ctx->get_constant_column(1));
+        }
+        int32_t counter_num;
+        if (ctx->get_num_args() > 2) {
+            counter_num = ColumnHelper::get_const_value<TYPE_INT>(ctx->get_constant_column(2));
+        } else {
+            counter_num = std::min(std::max(2 * k, 100), MAX_COUNTER_NUM);
+        }
+        DCHECK_GT(k, 0);
+        DCHECK_GT(counter_num, 0);
+        DCHECK_LE(k, MAX_COUNTER_NUM);
+        DCHECK_LE(counter_num, MAX_COUNTER_NUM);
+        DCHECK_GE(counter_num, k);
+        return std::make_pair(k, counter_num);
+    }
+
+    void init_state_if_necessary(FunctionContext* ctx, AggDataPtr __restrict state) const {
+        if (this->data(state).is_init) {
+            return;
+        }
+        this->data(state).is_init = true;
+        const auto kv = get_k_and_counter_num(ctx);
+        this->data(state).reset(kv.first, kv.second);
+    }
+
+    void process_null(FunctionContext* ctx, AggDataPtr __restrict state) const {
+        init_state_if_necessary(ctx, state);
+        this->data(state).process_null(1);
+    }
+
+    void merge(FunctionContext* ctx, const Column* column, AggDataPtr state, size_t row_num) const override {
+        DCHECK(column->is_binary());
```

Comment:
would it be nullable(binary)?

Reply:
No, it won't be, defined by the definiation of the function.